firstuseauthenticator/firstuseauthenticator.py

```diff
@@ -147,14 +147,18 @@ def authenticate(self, handler, data):
         password = data['password']
         # Don't enforce password length requirement on existing users, since that can
         # lock users out of their hubs.
-        if not self._validate_password(password) and not self._user_exists(username):
-            handler.custom_login_error = (
-                'Password too short! Please choose a password at least %d characters long.'
-                % self.min_password_length
-            )
-
-            self.log.error(handler.custom_login_error)
-            return None
+        with dbm.open(self.dbm_path, 'c', 0o600) as db:
+            stored_pw = db.get(username.encode(), None)
+            #check if there is no password for the user
+            # check if the password is adhering to the constraint
+            if stored_pw is None and not self._validate_password(password):
```

Comment:
This looks super good @saisiddhant12 :+1:

One more small nit before merging this:
These steps (opening the db and checking if the user's password was set) are already happening [here](https://github.com/jupyterhub/firstuseauthenticator/blob/master/firstuseauthenticator/firstuseauthenticator.py#L158-L165).
We should move the validate password part inside that check so we don't duplicate code too much.
What do you think?

Reply:
Thanks @GeorgianaElena , 
I was also thinking on same lines minutes ago, because lets say someone makes some changes/enhances it  in future, there should be a single place where he/she would do the change

the function definition would change to something like this 
```
    def _validate_password(self, password):
        with dbm.open(self.dbm_path, 'c', 0o600) as db:
            stored_pw = db.get(username.encode(), None)
            if ((len(password) >= self.min_password_length) 
                and (stored_pw is not None)):
                return stored_pw
            else:
                return False

Reply:
If we move the entire logic inside a function we'll still have to open the database twice to check the user's password during the authentication. First time when we're checking the password's length and the second time when we're checking if the user types the correct password [here](https://github.com/jupyterhub/firstuseauthenticator/blob/master/firstuseauthenticator/firstuseauthenticator.py#L158-L163).

To open the database just once, we could move the password validation part:
```python
if not self._validate_password(password):
    handler.custom_login_error = (
        'Password too short! Please choose a password at least %d characters long.'
        % self.min_password_length
     )
     self.log.error(handler.custom_login_error)
     return None
```

inside the `else` branch here, just before an entry in database for the user and their password is added: https://github.com/jupyterhub/firstuseauthenticator/blob/be01334bbd888a32ad330eb1fe4ba33ac9ece9ed/firstuseauthenticator/firstuseauthenticator.py#L163-L165

Reply:
i was thinking to call  the function once and store the returned value in a variable ?
`self._password_holder = self._validate_password(password)`
what are your thoughts on this ?

Reply:
The idea is it will be called once and stored in variable and the variable will be used again and again instead of function call. Do you think this will do the job?

Reply:
Ahhh, now i got your intention, basically, you are asking if we can combine both the the password validation into single one
Yes I will do the same and commit it